addons/web/static/src/js/core/owl_popover.js

```diff
@@ -0,0 +1,257 @@
+                    top: targetTop + targetHeight,
+                    left: targetLeft - (popoverWidth - targetWidth) / 2,
+                },
+                left: {
+                    name: 'left',
+                    top: targetTop - (popoverHeight - targetHeight) / 2,
+                    left: targetLeft - popoverWidth,
+                },
+            };
+        }
+
+        /**
+         * Check if the element is in the viewport.
+         * @private
+         * @static
+         * @param {HTMLElement}
+         * @returns {Boolean} True if the element currently fits inside 
+         *                    the viewport, false otherwise.
+         */
+        static _isInViewport(element) {
+            const rect = element.getBoundingClientRect();
+            const html = document.documentElement;
+            return (
+                rect.top >= 0 &&
+                rect.left >= 0 &&
+                rect.bottom <= (window.innerHeight || html.clientHeight) &&
+                rect.right <= (window.innerWidth || html.clientWidth)
+            );
+        }
+
+        /**
+         * Hide any displayed popover, display the popover in argument and mark
+         * it as the one being displayed in the class attribute. Add an event
+         * listener on the document to detect any click outside the popover as a
+         * closing event.
+         * Note that the event listener is added for the capture phase, meaning
+         * that it will *always* run before the click listener set on a popover
+         * template. This means that there can be no race condition regarding
+         * clicking on another popover trigger: any existing popover will be
+         * removed during the capture phase, then a new listener will be added
+         * in the bubbling phase of the new popover' trigger click event.
+         * @static
+         * @param {Popover} popover The popover component to display.
+         */
+        static display(popover) {
+            /* this should never happen because of the way the events are handled,
+            but in case an imaginative dev puts a popover inside another one, i'd
+            prefer the first one to be hidden anyway, causing the second one to
+            never display itself (since its target won't be in the viewport) */
+            if (this.displayed) {
+                this.displayed._close();
+            }
+            if (!this.isListening) {
+                document.addEventListener('click', Popover.documentClickHandler, { capture: true });
+                this.isListening = true;
+            }
+            this.displayed = popover;
+            popover.state.displayed = true;
+        }
+
+        /**
+         * Hide any displayed popover and remove event listeners.
+         */
+        static hide() {
+            const popover = this.displayed;
+            document.removeEventListener('click', Popover.documentClickHandler, { capture: true });
```

Comment:
So:
** when there is no popover **
there is a click on a target. 
The class starts listening on click, and a popover is displayed

**There is already a popover**
The old popover is destroyed first, because the class listener is in the capture phase
The class stops listening on click with hide()
The class starts listening on click with display()
The new popover is displayed

Is that right ?
Then I guess the class should not stop and restart, but rather stay in listening mode in that case (I know this is just one listener, but the case where we want to close a popover and reopen one is fairly frequent).
I don't know if this remark is relevant in terms of performance. nor if it is doable.....

Reply:
it's completely relevant (removing a listener to add it back seems stupid), I just didn't find a way to do it cleanly.

I'd need to know in advance in the `click` target is the trigger of another popover. Perhaps I should simply add a `data-popover` attribute on the outer div of the `Popover` component template and check for that and only remove the listener if the event's target does not have such an ancestor using [`Element.closest`](https://developer.mozilla.org/en-US/docs/Web/API/Element/closest)to find out? But it might be worse from a perf pov than removing/adding a listener, not familiar with the perfs of both these calls...

Reply:
that's why there an `isListening` attribute on the class btw, i wanted to toggle listening on/off only if it didn't match the expected state but i should remove that since adding the same listener multiple times does nothing, and removing a listener that isn't there does nothing